### PR TITLE
Honor the editor configured in the config

### PR DESCRIPTION
**Pull request by romanofski**
Always pick the editor configured in notmuch if EDITOR is unset.

Issue: #50

### CODE REVIEW

**Comment by frasertweedale** [comment]
On Thu, Aug 31, 2017 at 01:05:07AM +0000, Roman Joost wrote:
> @romanofski requested your review on: purebred-mua/purebred#52 Honor the editor configured in the config.
> 
Thx, I'll review tonight or maybe tomorrow.

**Comment by romanofski** [comment]
Will do!

**Comment by romanofski** [comment]
Ugh... right against the wall.

I've got:

```haskell
data NotmuchSettings m = NotmuchSettings
    { _nmSearch :: T.Text
    , _nmDatabase :: m FilePath
    , _nmNewTag :: T.Text
    }

nmDatabase :: Lens' (NotmuchSettings m) (m FilePath)
nmDatabase = lens _nmDatabase (\nms fp -> nms { _nmDatabase = fp })

type UserConfiguration = Configuration IO
type InternalConfiguration = Configuration Identity
```
So far so good. But going from UserConfiguration to InternalConfiguration is where I'm lost. How the hell do I go from IO -> Identity?

**Comment by frasertweedale** [comment]
@romanofski follow the types.  What can you do with the type of a lens (including a
composed lens)?

```haskell
type Lens s t a b = (a -> f b) -> s -> f t
```

Given the target type:

```haskell
processConfig :: Configuration IO -> Configuration Identity
```

How can you use a lens to reach the goal?

Let me know if you are still stuck after a while :)

**Comment by romanofski** [comment]
I've found a way to work my brain around a solution. Not sure if it's convoluted, but type checks:

```Haskell
processConfig :: UserConfiguration -> IO InternalConfiguration
processConfig cfg = do
  fp <- view (confNotmuch . nmDatabase) cfg
  pure $ set (confNotmuch . nmDatabase) (pure fp) cfg
```

**Comment by frasertweedale** [comment]
@romanofski well done.  Isn't type tetris fun!

Type golf is even better :)

```haskell
processConfig :: UserConfiguration -> IO InternalConfiguration
processConfig = (confNotmuch . nmDatabase) (fmap pure)
```

I often find it helps to think of an optic (lenses in particular) like `traverse` - gives you a way of
propagating an effect at the focus of the lens, out to the outside of the structure.  In the previous code we had:

```haskell
processConfig :: Configuration (IO FilePath) -> IO (Configuration FilePath)
processConfig = (confNotmuch . nmDatabase) id
```

So above, `id` give us `IO FilePath` - the `IO` gets pushed to the "outside" and we keep `FilePath` as the type at the focus.

Now in the new code:
```haskell
processConfig :: Configuration IO -> IO (Configuration Identity)
processConfig = (confNotmuch . nmDatabase) (fmap pure)
```
we have `fmap pure :: IO FilePath -> IO (Identity FilePath)`.  Again the `IO` gets propagated out,
and we keep `Identity FilePath` as the type at the focus, which is precisely what we need.

HTH!

**Comment by romanofski** [comment]
Cheers for the explanation. Had to quickly push my changes, so still working on it :)

File: src/Config/Main.hs [hunk 1]
```diff
@@ -45,7 +46,7 @@ defaultConfig =
       , _nmDatabase = getDatabasePath
       , _nmNewTag = "unread"
       }
-    , _confEditor = "vi"
+    , _confEditor = lookupEnv "EDITOR" >>= pure . fromMaybe "vi"
```
Comment (frasertweedale) [inline src/Config/Main.hs]: This can be `fromMaybe "vi" <$> lookupEnv "EDITOR"`

**Comment by frasertweedale** [comment]
For reading the `Identity`-wrapped fields, have a look at `Control.Lens.Wrapped`.
Lets you get at the thing inside a "wrapping" type, which includes `Identity`.

```
λ> import Data.Functor.Identity (Identity(..))
λ> import Control.Lens (view, _Wrapped)
λ> view _Wrapped (Identity "hello")
"hello"
```

Then instead of importing `runIdentity` and applying it to the result of a `view`, you
can compose the `_Wrapped` iso into the optic, e.g.:

```haskell
  let editor = view (asConfig . confEditor . _Wrapped) s 
```

**Comment by romanofski** [comment]
Ah cheers for the hint.

I'm a bit puzzled that I couldn't repeat my luck I had before by "dumbing" it down a bit. So again I have:

```Haskell
processConfig :: UserConfiguration -> IO InternalConfiguration
processConfig cfg = do
    fp <- view (confNotmuch . nmDatabase) cfg
    editor <- view confEditor cfg
    let cfg' =
            set (confNotmuch . nmDatabase) (pure fp) cfg &
            set confEditor (pure editor)
    pure cfg'
```
but I get the error:

```
    • Couldn't match type ‘IO’ with ‘Data.Functor.Identity.Identity’
      Expected type: IO InternalConfiguration
        Actual type: IO (Configuration IO)
    • In a stmt of a 'do' block: pure cfg'
      In the expression:
        do { fp <- view (confNotmuch . nmDatabase) cfg;
             editor <- view confEditor cfg;
             let cfg'
                   = set (confNotmuch . nmDatabase) (pure fp) cfg
                     & set confEditor (pure editor);
             pure cfg' }
      In an equation for ‘processConfig’:
          processConfig cfg
            = do { fp <- view (confNotmuch . nmDatabase) cfg;
                   editor <- view confEditor cfg;
                   let cfg' = ...;
                   .... }
```
which I'm surprised about. I'm "pulling" the String out of the IO computation so my reasoning is, that I should be able to "set" it back into the config, but it seems the compiler still assuming there is still an IO type there?

**Comment by frasertweedale** [comment]
I think you want something like:

```haskell
processConfig :: UserConfiguration -> IO InternalConfiguration
processConfig cfg = do
    fp <- view (confNotmuch . nmDatabase) cfg
    editor <- view confEditor cfg
    pure $ cfg
      & set (confNotmuch . nmDatabase) (pure fp)
      & set confEditor (pure editor)
```

**Comment by romanofski** [comment]
Hm.. nope. Computer says: "no".

**Comment by frasertweedale** [comment]
@romanofski uh, so it turns out we need just as many type variables after all, otherwise
we get type mismatches during the "transitional" phase when processing multiple fields.
I'm undecided which approach is worst... the old way we were doing it or the new way.  But I am leaning towards the OLD way now.

FWIW here is (an example of) what you'd have to do under the current paradigm:
```haskell
data Config' m m1 = Config (m String) (m1 String)
type Config m = Config' m m
type UserConfig = Config IO
type InternalConfig = Config Identity

db :: Lens (Config' m m1) (Config' m' m1) (m String) (m' String)
db f (Config a b) = fmap (\a' -> Config a' b) (f a)

ed :: Lens (Config' m m1) (Config' m m1') (m1 String) (m1' String)
ed f (Config a b) = fmap (\b' -> Config a b') (f b)

processConfig :: UserConfig -> IO InternalConfig
processConfig = db (fmap pure) >=> ed (fmap pure)
```

I really think I prefer the old way, which would be:
```
data Config a b = Config a b
type UserConfig = Config (IO String) (IO String)
type InternalConfig = Config String String

db :: Lens (Config a b) (Config a' b) a a'
db f (Config a b) = fmap (\a' -> Config a' b) (f a)

ed :: Lens (Config a b) (Config a b') b b'
ed f (Config a b) = fmap (\b' -> Config a b') (f b)

processConfig :: UserConfig -> IO InternalConfig
processConfig = db id >=> ed id
```

Mainly because if we need all the typevars anyway, I'd rather have them correspond
to inhabitable type (just seems "easier to understand" that way).  It also allows to
vary types between UserConfig and InternalConfig more, and we can do away with
the `_Wrapped` business that was necessary when abstracting over `(* -> *)` and using
`Identity` for the internal config type.

So... sorry for taking us down the garden path on this one, but I think we should change back :)

**Comment by romanofski** [comment]
Yeah I'd also opt for an easier to understand approach. At least this doesn't seem to warrant the complexity. No worries, I'll change it back. Not today anymore tho.

**Comment by romanofski** [comment]
Still can't make the type system happy. Would you mind to check my `confNotmuch` lens if it might create the error I run into in `Purebred.hs:105:9`?

```
    • Couldn't match type ‘[Char]’ with ‘IO FilePath’
        arising from a functional dependency between:
          constraint ‘mtl-2.2.1:Control.Monad.Reader.Class.MonadReader
                        (Configuration (IO (IO FilePath)) (IO String))
                        ((->) UserConfiguration)’
            arising from a use of ‘view’
          instance ‘mtl-2.2.1:Control.Monad.Reader.Class.MonadReader
                      r ((->) r)’
            at <no location info>
    • In a stmt of a 'do' block:
        fp <- view (confNotmuch . nmDatabase) cfg
      In the expression:
        do { ed <- view confEditor cfg;
             fp <- view (confNotmuch . nmDatabase) cfg;
             pure
             $ cfg & set (confNotmuch . nmDatabase) fp & set confEditor ed }
      In an equation for ‘processConfig’:
          processConfig cfg
            = do { ed <- view confEditor cfg;
                   fp <- view (confNotmuch . nmDatabase) cfg;
                   pure
                   $ cfg & set (confNotmuch . nmDatabase) fp & set confEditor ed }
```

Is it me who's not seeing it or maybe my botched lens?

**Comment by frasertweedale** [comment]
looks like somehow you got an `IO (IO Filepath)` in there.  Feel free to commit WIP and I'll take a look at it.

**Comment by romanofski** [comment]
Latest WIP is already up, but let me give it another go to figure out what's wrong. Maybe I find the issue ...

**Comment by romanofski** [comment]
Oohh I've figured it out. W00t. Will push asap...

**Comment by romanofski** [comment]
Just to write it down for myself: I've made a mistake with the lens and it's type signature. While it was type checking, I basically "mapped" it to the wrong types. This ended either up picking the "editor" (IO String) or a double IO Monad. Splitting up the first `view (confNotmuch . nmDatabase) cfg` into smaller pieces helped to figure out what was wrong.

File: src/Config/Main.hs (hunk 1 above)
Comment (frasertweedale) [reply src/Config/Main.hs]: ... this simplification is still applicable ;)

File: src/Purebred.hs [hunk 2]
```diff
@@ -99,7 +100,10 @@ launch cfg = do
     void $ defaultMain (theApp s) s
 
 processConfig :: UserConfiguration -> IO InternalConfiguration
-processConfig = (confNotmuch . nmDatabase) id
+processConfig cfg = do
+    fp <- view (confNotmuch . nmDatabase) cfg
+    ed <- view confEditor cfg
+    pure $ cfg & set (confNotmuch . nmDatabase) (fp) & set confEditor (ed)
```
Comment (frasertweedale) [inline src/Purebred.hs]: would be good to break this across multiple lines & remove extraneous parens, e.g.
```haskell
pure $ cfg
  & set (confNotmuch . nmDatabase) fp
  & set confEditor ed
```

Comment (romanofski) [reply src/Purebred.hs]: Ah dang! Yes of course. Will do.

**Pull request merged**